File: plsql-frontend/src/main/java/br/com/felipezorzo/sonar/plsql/squid/PlSqlAstScanner.java
/*
 * Sonar PL/SQL Plugin (Community)
 * Copyright (C) 2015 Felipe Zorzo
 * felipe.b.zorzo@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package br.com.felipezorzo.sonar.plsql.squid;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.Collection;

import org.sonar.squidbridge.AstScanner;
import org.sonar.squidbridge.AstScanner.Builder;
import org.sonar.squidbridge.SquidAstVisitor;
import org.sonar.squidbridge.SquidAstVisitorContextImpl;
import org.sonar.squidbridge.api.SourceCode;
import org.sonar.squidbridge.api.SourceFile;
import org.sonar.squidbridge.api.SourceProject;
import org.sonar.squidbridge.indexer.QueryByType;
import org.sonar.squidbridge.metrics.CommentsVisitor;
import org.sonar.squidbridge.metrics.ComplexityVisitor;
import org.sonar.squidbridge.metrics.LinesVisitor;

import com.google.common.collect.ImmutableList;
import com.sonar.sslr.api.AstNodeType;
import com.sonar.sslr.api.Grammar;
import com.sonar.sslr.impl.Parser;

import br.com.felipezorzo.sonar.plsql.api.PlSqlGrammar;
import br.com.felipezorzo.sonar.plsql.api.PlSqlKeyword;
import br.com.felipezorzo.sonar.plsql.api.PlSqlMetric;
import br.com.felipezorzo.sonar.plsql.parser.PlSqlParser;

public class PlSqlAstScanner {

    private PlSqlAstScanner() {
    }
    
    public static SourceFile scanSingleFile(File file, SquidAstVisitor<Grammar> visitor) {
        return scanSingleFile(file, ImmutableList.of(visitor));
    }

    public static SourceFile scanSingleFile(File file, Collection<SquidAstVisitor<Grammar>> visitors) {
        if (!file.isFile()) {
            throw new IllegalArgumentException("File '" + file + "' not found.");
        }
        AstScanner<Grammar> scanner = create(new PlSqlConfiguration(StandardCharsets.UTF_8), visitors);
        scanner.scanFile(file);
        Collection<SourceCode> sources = scanner.getIndex().search(new QueryByType(SourceFile.class));
        if (sources.size() != 1) {
            throw new IllegalStateException("Only one SourceFile was expected whereas "
                            + sources.size() + " has been returned.");
        }
        return (SourceFile) sources.iterator().next();
    }
    
    public static AstScanner<Grammar> create(PlSqlConfiguration conf, Collection<SquidAstVisitor<Grammar>> visitors) {
        final SquidAstVisitorContextImpl<Grammar> context = 
                new SquidAstVisitorContextImpl<>(new SourceProject("PL/SQL Project"));
        final Parser<Grammar> parser = PlSqlParser.create(conf);

        AstScanner.Builder<Grammar> builder = AstScanner.<Grammar>builder(
                context).setBaseParser(parser);
        
        builder.setFilesMetric(PlSqlMetric.FILES);
        setCommentAnalyser(builder);
        setMetrics(conf, builder);

        /* External visitors (typically Check ones) */
        for (SquidAstVisitor<Grammar> visitor : visitors) {
            if (visitor instanceof CharsetAwareVisitor) {
                ((CharsetAwareVisitor) visitor).setCharset(conf.getCharset());
            }
            builder.withSquidAstVisitor(visitor);
        }

        return builder.build();
    }
    
    private static void setMetrics(PlSqlConfiguration conf, Builder<Grammar> builder) {
        builder.withSquidAstVisitor(new LinesVisitor<>(PlSqlMetric.LINES));
        builder.withSquidAstVisitor(new PlSqlLinesOfCodeVisitor(PlSqlMetric.LINES_OF_CODE));
        builder.withSquidAstVisitor(CommentsVisitor.<Grammar>builder().withCommentMetric(PlSqlMetric.COMMENT_LINES)
                .withNoSonar(true)
                .withIgnoreHeaderComment(conf.getIgnoreHeaderComments())
                .build());
        
        AstNodeType[] complexityAstNodeType = new AstNodeType[] {
                PlSqlGrammar.CREATE_PROCEDURE,
                PlSqlGrammar.CREATE_FUNCTION,
                PlSqlGrammar.ANONYMOUS_BLOCK,
                
                PlSqlGrammar.PROCEDURE_DECLARATION,
                PlSqlGrammar.FUNCTION_DECLARATION,
                
                PlSqlGrammar.LOOP_STATEMENT,
                PlSqlGrammar.CONTINUE_STATEMENT,
                PlSqlGrammar.FOR_STATEMENT,
                PlSqlGrammar.EXIT_STATEMENT,
                PlSqlGrammar.IF_STATEMENT,
                PlSqlGrammar.RAISE_STATEMENT,
                PlSqlGrammar.RETURN_STATEMENT,
                PlSqlGrammar.WHILE_STATEMENT,
                
                // this includes WHEN in exception handlers, exit/continue statements and CASE expressions
                PlSqlKeyword.WHEN,
                PlSqlKeyword.ELSIF
        };
        builder.withSquidAstVisitor(ComplexityVisitor.<Grammar> builder().setMetricDef(PlSqlMetric.COMPLEXITY)
                .subscribeTo(complexityAstNodeType).build());
    }

    private static void setCommentAnalyser(AstScanner.Builder<Grammar> builder) {
        builder.setCommentAnalyser(new PlSqlCommentAnalyzer());
    }
}

File: sonar-zpa-plugin/src/main/kotlin/com/felipebz/zpa/squid/PlSqlAstScanner.kt
/**
 * Z PL/SQL Analyzer
 * Copyright (C) 2015-2025 Felipe Zorzo
 * mailto:felipe AT felipezorzo DOT com DOT br
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.felipebz.zpa.squid

import org.sonar.api.batch.fs.InputFile
import org.sonar.api.batch.fs.TextRange
import org.sonar.api.batch.measure.Metric
import org.sonar.api.batch.sensor.SensorContext
import org.sonar.api.batch.sensor.issue.NewIssue
import org.sonar.api.batch.sensor.issue.NewIssueLocation
import org.sonar.api.issue.NoSonarFilter
import org.sonar.api.measures.CoreMetrics
import org.sonar.api.measures.FileLinesContextFactory
import com.felipebz.zpa.PlSqlChecks
import com.felipebz.zpa.checks.IssueLocation
import com.felipebz.zpa.highlight.PlSqlHighlighterVisitor
import com.felipebz.zpa.metadata.FormsMetadata
import com.felipebz.zpa.metrics.CpdVisitor
import com.felipebz.zpa.rules.SonarQubeRuleKeyAdapter
import com.felipebz.zpa.symbols.ObjectLocator
import com.felipebz.zpa.symbols.SonarQubeSymbolTable
import com.felipebz.zpa.api.PlSqlFile
import com.felipebz.zpa.api.checks.PlSqlVisitor
import java.io.Serializable
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class PlSqlAstScanner(private val context: SensorContext,
                      checks: Collection<PlSqlVisitor>,
                      private val noSonarFilter: NoSonarFilter,
                      formsMetadata: FormsMetadata?,
                      isErrorRecoveryEnabled: Boolean,
                      private val fileLinesContextFactory: FileLinesContextFactory?) {

    private val astScanner: AstScanner =
        AstScanner(checks, formsMetadata, isErrorRecoveryEnabled, context.fileSystem().encoding())

    private lateinit var plsqlChecks: PlSqlChecks

    constructor(context: SensorContext,
                checks: PlSqlChecks,
                noSonarFilter: NoSonarFilter,
                formsMetadata: FormsMetadata?,
                isErrorRecoveryEnabled: Boolean,
                fileLinesContextFactory: FileLinesContextFactory,
                objectLocator: ObjectLocator) : this(context, checks.all(), noSonarFilter, formsMetadata, isErrorRecoveryEnabled, fileLinesContextFactory) {
        this.plsqlChecks = checks
        objectLocator.setScope(astScanner.globalScope)
    }

    fun scanFile(inputFile: InputFile) {
        val plSqlFile = SonarQubePlSqlFile(inputFile)
        if (plSqlFile.type() == PlSqlFile.Type.MAIN) {
            scanMainFile(plSqlFile)
        } else {
            scanTestFile(plSqlFile)
        }
    }

    private fun scanMainFile(plSqlFile: SonarQubePlSqlFile) {
        println("Scanning file ${plSqlFile.fileName()}")
        val inputFile = plSqlFile.inputFile
        val result = try {
            astScanner.scanFile(
                plSqlFile,
                listOf(PlSqlHighlighterVisitor(context, inputFile), CpdVisitor(context, inputFile))
            )
        } catch (e: Exception) {
            println("Error scanning file ${plSqlFile.fileName()}: ${e.message}")
            e.printStackTrace()
            throw e
        }
        println("File ${plSqlFile.fileName()} scanned")

        noSonarFilter.noSonarInFile(inputFile, result.linesWithNoSonar)

        lock.withLock { // needed because SonarQube 7.6 save this data in a non thread-safe way
            saveIssues(inputFile, result.issues)

            val symbolSaver = SonarQubeSymbolTable(context, inputFile)
            symbolSaver.save(result.symbols)

            saveMetricOnFile(inputFile, CoreMetrics.STATEMENTS, result.numberOfStatements)
            saveMetricOnFile(inputFile, CoreMetrics.NCLOC, result.linesOfCode)
            println("NCLOC: ${result.linesOfCode}")
            saveMetricOnFile(inputFile, CoreMetrics.COMMENT_LINES, result.linesOfComments)
            saveMetricOnFile(inputFile, CoreMetrics.COMPLEXITY, result.complexity)
            saveMetricOnFile(inputFile, CoreMetrics.FUNCTIONS, result.numberOfFunctions)


            if (fileLinesContextFactory != null) {
                val fileLinesContext = fileLinesContextFactory.createFor(inputFile)
                for (line in result.executableLines) {
                    fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, 1)
                }
                fileLinesContext.save()
            }
        }
    }

    private fun scanTestFile(plSqlFile: SonarQubePlSqlFile) {
        val inputFile = plSqlFile.inputFile
        val result = astScanner.scanFile(plSqlFile, listOf(PlSqlHighlighterVisitor(context, inputFile)))

        noSonarFilter.noSonarInFile(inputFile, result.linesWithNoSonar)

        lock.withLock { // needed because SonarQube 7.6 save this data in a non thread-safe way
            saveIssues(inputFile, result.issues)

            val symbolTable = SonarQubeSymbolTable(context, inputFile)
            symbolTable.save(result.symbols)
        }
    }

    private fun saveIssues(inputFile: InputFile, issues: List<ZpaIssue>) {
        for (issue in issues) {
            val rule = plsqlChecks.ruleKey(issue.check) as SonarQubeRuleKeyAdapter

            val newIssue = context.newIssue().forRule(rule.ruleKey)

            val cost = issue.cost
            if (cost != null) {
                newIssue.gap(cost.toDouble())
            }

            newIssue.at(newLocation(inputFile, newIssue, issue.primaryLocation))

            for (secondaryLocation in issue.secondaryLocations) {
                newIssue.addLocation(newLocation(inputFile, newIssue, secondaryLocation))
            }

            newIssue.save()
        }
    }

    private fun <T : Serializable> saveMetricOnFile(inputFile: InputFile, metric: Metric<T>, value: T) {
        context.newMeasure<T>()
                .on(inputFile)
                .forMetric(metric)
                .withValue(value)
                .save()
    }

    private fun newLocation(inputFile: InputFile, issue: NewIssue, location: IssueLocation): NewIssueLocation {
        val newLocation = issue.newLocation().on(inputFile)
        if (location.startLine() != IssueLocation.UNDEFINED_LINE) {
            val range: TextRange = if (location.startLineOffset() == IssueLocation.UNDEFINED_OFFSET) {
                inputFile.selectLine(location.startLine())
            } else {
                inputFile.newRange(location.startLine(), location.startLineOffset(), location.endLine(),
                        location.endLineOffset())
            }
            newLocation.at(range)
        }

        newLocation.message(location.message())
        return newLocation
    }

    companion object {
        private val lock = ReentrantLock()
    }

}
